dace/transformation/dataflow/otf_map_fusion.py

```diff
@@ -96,102 +144,185 @@ def apply(self, graph: SDFGState, sdfg: SDFG):
             self.second_map_entry.remove_out_connector(edge.src_conn)
             graph.remove_edge(edge)
 
-        # OTF: Replace read memlet by copying full map content of corresponding write memlet
-        for array in read_memlets:
-            write_memlet = write_memlets[array]
-            write_accesses = write_memlet.subset.ranges
-            for read_accesses in read_memlets[array]:
-                # Map read access to write access
-                param_subs = OTFMapFusion.solve(first_map_entry.map.params, write_accesses,
-                                                self.second_map_entry.map.params, read_accesses)
-
-                new_nodes = self._copy_first_map_contents(sdfg, graph, first_map_entry)
-
+        # Phase 3: OTF - copy content for each in memlet of second map
+        for array in in_memlets:
+            out_memlet = out_memlets[array]
+            out_accesses = tuple(out_memlet.subset.ranges)
+            for in_accesses in in_memlets[array]:
+                # Add intermediate scalar for output of copied map content
                 tmp_name = "__otf"
-                tmp_name, _ = sdfg.add_scalar(tmp_name, sdfg.arrays[array].dtype, transient=True, find_new_name=True)
+                tmp_name, _ = sdfg.add_scalar(tmp_name,
+                                              sdfg.arrays[array].dtype,
+                                              transient=True,
+                                              find_new_name=True,
+                                              lifetime=dtypes.AllocationLifetime.Scope)
                 tmp_access = graph.add_access(tmp_name)
+                tmp_access.setzero = True
 
-                # Connect read-in memlets to tmp_access node
-                read_memlet_group = read_memlets[array][read_accesses]
+                # Connect in memlets of second map to this scalar
+                read_memlet_group = in_memlets[array][in_accesses]
                 for edge in read_memlet_group:
                     graph.add_edge(tmp_access, None, edge.dst, edge.dst_conn, Memlet(tmp_name))
 
-                # Connect new sub graph
-                for node in new_nodes:
+                # Solve index mapping between memlets
+                param_mapping = OTFMapFusion.solve(first_map_entry.map.params, out_accesses,
+                                                   self.second_map_entry.map.params, in_accesses)
+
+                # Copy actual content
+                first_map_inner_nodes = self._copy_first_map_contents(sdfg, graph, first_map_entry)
+
+                # Prepare: Symbol collision
+                # Second map defines symbols (params) and becomes the new outer scope of the copied nodes.
+                # Thus, newly introduced symbols in this subgraph must be renamed first to avoid conflicts.
+                # Note symbols of first map must be re-named according to the param mapping.
+                for node in first_map_inner_nodes:
```

Comment:
this won't work for multi-level nested maps, right?

Reply:
No, basically everyhting nested, also no states, nested SDFGs etc. Here I would need your help to generalize it. There must be a helper function which could replace symbols in subgraphs?

Reply:
`SubgraphView.replace` is your friend